src/RoyalLibrary/SortingExtensions.cs

```diff
@@ -0,0 +1,156 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+namespace ByteDecoder.RoyalLibrary
+{
+
+    /// <summary>
+    /// Advanced sorting for collections
+    /// </summary>
+    public static class SortingExtensions
+    {
+
+
+        /// <summary>
+        /// returns a new sorted ICollection 
+        /// </summary>
+        /// <param name="source"></param>
+        /// <returns>An Enumerable sorted collection by LastName</returns>
+        public async static Task<ICollection<string>> RoyalSortLastAsync(this IEnumerable<string> source)
+        {
+            int rowsCount = source.Count();
+            byte lastspaceIndex = 0;
+            string firstName, lastName = null;
+            var tempDic = new Dictionary<string, string>(rowsCount);
+            var tempList = new List<string>(rowsCount);
+            var tempResult = new List<string>(rowsCount);
+
+            await Task.Run(() =>
+            {
+                foreach (var item in source)
+                {
+
+                    var spaces = item.Count(char.IsWhiteSpace);
+                    lastspaceIndex = (byte)item.IndexOfNth(' ', spaces - 1);
+
+                    //get the whole firstname(s) , then the lastName, add them to tempDic which will be used
+                    //to get firstname(s) by lastName
+                    //lastname will be added to the temp list and sorted everytime 
+                    firstName = item.Substring(0, lastspaceIndex);
+                    var lastNameLength = item.Length - firstName.Length;
+                    lastName = item.Substring(lastspaceIndex, lastNameLength);
+
+                    tempDic.Add(firstName, lastName);

```

Comment:
There is an error here. Since you are using a dictionary and you already have the key *Ai Bi* and there is another entry with the same first name *Ai Bi* throws an error:

An exception of type 'System.ArgumentException' occurred in System.Private.CoreLib.dll but was not handled in user code: 'An item with the same key has already been added. Key: Ai Bi'

Reply:
Thanks for replying , I forgot duplicated keys with dictionary issue , now I resolved this and works well.
I have already committed it and test file executed successfully .